### PR TITLE
feat/718: add OpenAI Text to Speech as a Provider

### LINKED ISSUE #718: Add OpenAI as a Provider for Text to speech feature.
### Is your enhancement related to a problem? Please describe.

OpenAI provides a [Text to speech](https://platform.openai.com/docs/guides/text-to-speech)(TTS) API to turn text into lifelike spoken audio. It supports [multiple languages](https://platform.openai.com/docs/guides/text-to-speech/supported-languages) and is optimized for English. It could be a good alternative provider option to the current provider Azure AI Speech.

### Designs

_No response_

### Describe alternatives you've considered

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Sidsector9**
### Description of the Change

In addition to MS Azure Speech API, this PR adds OpenAI's Text to Speech service as a provider.

Closes #718 

### Changelog Entry

> Added - OpenAI Text to Speech as a Provider

### Credits

Props @iamdharmesh @Sidsector9 

### Checklist:

- [ ] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

### CODE REVIEW

File: includes/Classifai/Providers/OpenAI/TextToSpeech.php [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+			$this->feature_instance->get_option_name() . '_section',
+			[
+				'option_index'  => static::ID,
+				'label_for'     => 'tts_model',
+				'options'       => [
+					'tts-1'    => __( 'Text-to-speech 1 (Optimized for speed)', 'classifai' ),
+					'tts-1-hd' => __( 'Text-to-speech 1 HD (Optimized for quality)', 'classifai' ),
+				],
+				'default_value' => $settings['tts_model'],
+				'description'   => __( 'Select a model depending on your requirements.', 'classifai' ),
```
Comment (dkotter) [inline includes/Classifai/Providers/OpenAI/TextToSpeech.php]: Might be nice to link out to the [documentation](https://platform.openai.com/docs/models/tts) here so people can better determine the differences between the models.

Maybe just link the word `model` in that sentence to the docs?

File: includes/Classifai/Providers/OpenAI/TextToSpeech.php [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+				'options'       => [
+					'alloy'   => __( 'Alloy (male)', 'classifai' ),
+					'echo'    => __( 'Echo (male)', 'classifai' ),
+					'fable'   => __( 'Fable (male)', 'classifai' ),
+					'onyx'    => __( 'Onyx (male)', 'classifai' ),
+					'nova'    => __( 'Nova (female)', 'classifai' ),
+					'shimmer' => __( 'Shimmer (female)', 'classifai' ),
+				],
+				'default_value' => $settings['voice'],
+				'description'   => __( 'Select the desired speech voice.', 'classifai' ),
```
Comment (dkotter) [inline includes/Classifai/Providers/OpenAI/TextToSpeech.php]: Same thing here, might be nice to link to the [docs](https://platform.openai.com/docs/guides/text-to-speech) so people can listen to each voice.

File: includes/Classifai/Providers/OpenAI/TextToSpeech.php [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+					'aac'  => __( '.aac', 'classifai' ),
+					'flac' => __( '.flac', 'classifai' ),
+					'wav'  => __( '.wav', 'classifai' ),
+					'pcm'  => __( '.pcm', 'classifai' ),
+				],
+				'default_value' => $settings['format'],
+				'description'   => __( 'Select the desired audio format.', 'classifai' ),
+				'class'         => 'classifai-provider-field hidden provider-scope-' . static::ID,
+			]
+		);
```
Comment (dkotter) [inline includes/Classifai/Providers/OpenAI/TextToSpeech.php]: I think maybe we just get rid of this and hardcode to the `.mp3` format. I get that these are additional formats the API supports but not sure WordPress even supports all of these formats natively and for the sake of limiting options, I'd just remove this

File: includes/Classifai/Providers/OpenAI/TextToSpeech.php [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+		if ( ! current_user_can( 'edit_post', $post_id ) && ( ! defined( 'WP_CLI' ) || ! WP_CLI ) ) {
+			return new WP_Error(
+				'openai_text_to_speech_user_not_authorized',
+				esc_html__( 'Unauthorized user.', 'classifai' )
+			);
+		}
+
+		$feature             = new FeatureTextToSpeech();
+		$settings            = $feature->get_settings();
+		$post_content        = $feature->normalize_post_content( $post_id );
```
Comment (dkotter) [inline includes/Classifai/Providers/OpenAI/TextToSpeech.php:282]: Looks like there's a size limit on the content that we can send (4096 characters). Should we add some protections here that checks if we're above that limit and returns early with an error message?

File: includes/Classifai/Providers/OpenAI/TextToSpeech.php [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+			'method'  => 'POST',
+			'body'    => wp_json_encode( $request_body ),
+			'timeout' => 60, // phpcs:ignore WordPressVIPMinimum.Performance.RemoteRequestTimeout.timeout_timeout
+			'headers' => array(
+				'Authorization' => 'Bearer ' . $settings[ static::ID ]['api_key'],
+				'Content-Type'  => 'application/json',
+			),
+		);
+
+		$response = wp_remote_post( $this->api_url, $request_params );
```
Comment (dkotter) [inline includes/Classifai/Providers/OpenAI/TextToSpeech.php]: Any reason we couldn't use our `APIRequest` class here?

File: includes/Classifai/Providers/OpenAI/TextToSpeech.php [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+			return new WP_Error(
+				'openai_text_to_speech_unsuccessful_request',
+				esc_html__( 'HTTP request unsuccessful.', 'classifai' )
+			);
+		}
+
+		update_post_meta( $post_id, FeatureTextToSpeech::AUDIO_HASH_KEY, md5( $post_content ) );
+
+		return $response_body;
+	}
```
Comment (dkotter) [inline includes/Classifai/Providers/OpenAI/TextToSpeech.php:333]: Not crucial but I know our other Providers all have the `get_debug_information` method. Should we add that here as well?

**Review by dkotter — CHANGES_REQUESTED** [review]

File: includes/Classifai/Providers/OpenAI/TextToSpeech.php (hunk 4 above)
Comment (Sidsector9) [reply includes/Classifai/Providers/OpenAI/TextToSpeech.php:282]: Fixed in [45a2b26](https://github.com/10up/classifai/pull/747/commits/45a2b26040470dfaf084770fbbd2598d8ba4396f)

File: includes/Classifai/Providers/OpenAI/TextToSpeech.php (hunk 5 above)
Comment (Sidsector9) [reply includes/Classifai/Providers/OpenAI/TextToSpeech.php]: Fixed in [f637684](https://github.com/10up/classifai/pull/747/commits/f6376844d93d2c3f03aae904828cbee070c74a69)

File: includes/Classifai/Providers/OpenAI/TextToSpeech.php (hunk 6 above)
Comment (Sidsector9) [reply includes/Classifai/Providers/OpenAI/TextToSpeech.php:333]: Fixed in [416c03d](https://github.com/10up/classifai/pull/747/commits/416c03defdb295152603269de208304ebcb44d02)

File: includes/Classifai/Providers/OpenAI/TextToSpeech.php (hunk 3 above)
Comment (Sidsector9) [reply includes/Classifai/Providers/OpenAI/TextToSpeech.php]: I've limited the formats to `wav` and `mp3`.

**Review by dkotter — APPROVED** [review]

**Pull request merged**